Changes in module 'nf-core/merquryfk/merquryfk'
--- modules/nf-core/merquryfk/merquryfk/main.nf
+++ modules/nf-core/merquryfk/merquryfk/main.nf
@@ -41,8 +41,8 @@
     prefix = task.ext.prefix ?: "${meta.id}"
     def mat_ktab = matktab ? "${matktab.find{ it.toString().endsWith(".ktab") }}" : ''
     def pat_ktab = patktab ? "${patktab.find{ it.toString().endsWith(".ktab") }}" : ''
-    def FASTK_VERSION = 'f18a4e6d2207539f7b84461daebc54530a9559b0' // WARN: Version information not provided by tool on CLI. Please update this string when bumping container versions.
-    def MERQURY_VERSION = '8ae344092df5dcaf83cfb7f90f662597a9b1fc61' // WARN: Version information not provided by tool on CLI. Please update this string when bumping container versions.
+    def FASTK_VERSION = '666652151335353eef2fcd58880bcef5bc2928e1' // WARN: Version information not provided by tool on CLI. Please update this string when bumping container versions.
+    def MERQURY_VERSION = '666652151335353eef2fcd58880bcef5bc2928e1' // WARN: Version information not provided by tool on CLI. Please update this string when bumping container versions.
     """
     MerquryFK \\
         $args \\

************************************************************
