Changes in module 'nf-core/bedtools/bamtobed'
--- modules/nf-core/bedtools/bamtobed/main.nf
+++ modules/nf-core/bedtools/bamtobed/main.nf
@@ -25,7 +25,7 @@
         bamtobed \\
         $args \\
         -i $bam \\
-        | bedtools sort > ${prefix}.bed
+        > ${prefix}.bed
 
     cat <<-END_VERSIONS > versions.yml
     "${task.process}":

************************************************************
